### PR TITLE
[ENH] make ResNet layers dynamic via n_filters param

### PR DESCRIPTION
**Pull request by Rusheel86**
#### Reference Issues/PRs
Refs #9340

#### What does this implement/fix? Explain your changes.

Makes the number of residual blocks and key hyperparameters in `ResNetNetwork` dynamic via new exposed parameters:

- `n_filters` is a tuple of ints where `len(n_filters)` determines the number of residual blocks and each value sets the filter count for that block
- `kernel_sizes` is a tuple of ints controlling the kernel size of each of the 3 conv layers within every residual block
- Both parameters are exposed through `ResNetClassifier` and `ResNetRegressor` and passed down to `ResNetNetwork`
- Defaults `n_filters=(64, 128, 128)`, `kernel_sizes=(8, 5, 3)`, preserve the original paper behaviour (Wang et al. 2017)
- `build_network` refactored from 3 hardcoded copy-pasted blocks to a single loop

Also fixes a pre-existing bug where `ResNetRegressor` was passing `activation_hidden` as a kwarg to `ResNetNetwork.__init__`, which only accepts `activation`, causing a `TypeError` on instantiation.

#### Does your contribution introduce a new dependency? If yes, which one?
No.

#### What should a reviewer concentrate their feedback on?
- Correctness of the shortcut/residual connection logic inside the loop, specifically the channel expansion condition
-  Whether the exposed parameter interface (`n_filters`, `kernel_sizes`) is complete and appropriate

#### Did you add any tests for the change?
Updated `get_test_params` in `ResNetNetwork` to include a second param set with `n_filters=(32, 64)` to test a non-default block count.

#### Any other comments?
None.

##### For all contributions
- [x] I've added myself to the [[list of contributors](https://github.com/sktime/sktime/blob/main/CONTRIBUTORS.md)](https://github.com/sktime/sktime/blob/main/CONTRIBUTORS.md) with any new badges I've earned :-)
- [ ] Optionally, for added estimators: I've added myself and possibly to the `maintainers` tag
- [x] The PR title starts with either [ENH], [MNT], [DOC], or [BUG].

### CODE REVIEW

File: sktime/regression/deep_learning/resnet.py [hunk 1]
```diff
@@ -97,13 +101,15 @@ def __init__(
...
         self.use_bias = use_bias
         self.optimizer = optimizer
+        self.n_filters = n_filters
 
         super().__init__()
 
         self.history = None
         self._network = ResNetNetwork(
-            activation_hidden=self.activation_hidden,
+            activation=self.activation_hidden,
```
Comment (RecreationalMath) [inline sktime/regression/deep_learning/resnet.py:115]: Thanks for catching this bug.
Whenever you fix a bug, please ensure to mention it in your PR descrption or in a comment.

File: sktime/networks/resnet.py [hunk 2]
```diff
@@ -60,94 +64,44 @@ def build_network(self, input_shape, **kwargs):
...
-        output_block_3 = keras.layers.add([shortcut_y, conv_z])
-        output_block_3 = keras.layers.Activation(self.activation)(output_block_3)
+        x = input_layer
+
+        # stack residual blocks; number of blocks = len(self.n_filters)
+        for filters in self.n_filters:
+
+            # conv block: 3 conv layers with decreasing kernel sizes
+            conv_x = keras.layers.Conv1D(
+                filters=filters, kernel_size=8, padding="same"
```
Comment (RecreationalMath) [inline sktime/networks/resnet.py]: Similarly, `kernel_size` and `padding` should also be controlled via exposed parameters.

**Comment by Rusheel86** [comment]
Thanks for the detailed feedback @RecreationalMath, I've exposed `kernel_sizes` and `padding` as user-controllable parameters across `ResNetNetwork`, `ResNetClassifier`, and `ResNetRegressor`, with defaults matching the original paper.
I've also edited the pr description to mention the bug. 
Please let me know if any further changes are needed :)

File: sktime/classification/deep_learning/resnet.py [hunk 3]
```diff
@@ -45,6 +45,13 @@ class ResNetClassifier(BaseDeepClassifier):
         whether the layer uses a bias vector.
     optimizer : keras.optimizers object, default = Adam(lr=0.01)
         specify the optimizer and the learning rate to be used.
+    n_filters : tuple of int, default = (64, 128, 128)
+        Number of filters per residual block. Length determines number of
+        blocks.
+    kernel_sizes : tuple of int, default = (8, 5, 3)
+        Kernel sizes for the three conv layers in each residual block.
```
Comment (RecreationalMath) [inline sktime/classification/deep_learning/resnet.py]: I believe that having three conv layers is the default used in the referenced research paper, but apart from that, there is no limitation on the user to set it to three, right? 
So we should set the default as that, but the docstring should be more generic and describe the overall role of the parameter. something like.. `Tuple containing the kernel sizes for conv layers in each residual block, and the length of the tuple decides the number of conv layers to be used.` Do confirm with the research paper that this is the intended use of this parameter.

File: sktime/networks/resnet.py [hunk 4]
```diff
@@ -60,94 +77,44 @@ def build_network(self, input_shape, **kwargs):
...
-
-        output_block_3 = keras.layers.add([shortcut_y, conv_z])
-        output_block_3 = keras.layers.Activation(self.activation)(output_block_3)
+        x = input_layer
+
+        # stack residual blocks; number of blocks = len(self.n_filters)
+        for filters in self.n_filters:
+
+            # conv block: 3 conv layers with kernel sizes from self.kernel_sizes
+            conv_x = keras.layers.Conv1D(
```
Comment (RecreationalMath) [inline sktime/networks/resnet.py]: Corresponding to the change suggested in the docstring, the implementation here should also reflect that, i.e. the number of conv layers should be infered and not hardcoded.

File: sktime/classification/deep_learning/resnet.py (hunk 3 above)
Comment (Rusheel86) [reply sktime/classification/deep_learning/resnet.py]: Updated the docstring as suggested and checked the research paper. I've made it more generic to reflect that the length of the tuple now dictates the number of convolutional layers used in each block.

File: sktime/networks/resnet.py (hunk 4 above)
Comment (Rusheel86) [reply sktime/networks/resnet.py]: Yes, I've updated the code to infer the no. of conv layers.

**Comment by Rusheel86** [comment]
Thanks, all suggested changes have been addressed in the docstring of all 3 files and the number of conv layers are now inferred from `len(n_filters)`.

File: sktime/networks/resnet.py [hunk 5]
```diff
@@ -60,94 +78,35 @@ def build_network(self, input_shape, **kwargs):
...
+            for i, kernel_size in enumerate(self.kernel_sizes):
+                out = keras.layers.Conv1D(
+                    filters=filters, kernel_size=kernel_size, padding=self.padding
+                )(out)
+                out = keras.layers.BatchNormalization()(out)
+                if i < len(self.kernel_sizes) - 1:
+                    out = keras.layers.Activation(self.activation)(out)
+
+            # shortcut: expand channels if filter count changed, else batch norm only
+            if int(x.shape[-1]) == filters:
```
Comment (RecreationalMath) [inline sktime/networks/resnet.py:93]: There is an interesting divergence that happens here. 
Consider the scenario where the user passes a 64-variate time series (64 input channels) with the default n_filters=(64, 128, 128), then iteration 0 hits 64 == 64 and takes the BN-only path. The original code would always use Conv1D + BN for block 1, regardless of input channels.

In ResNet terminology, the original implementation used "projection shortcuts everywhere except the last block", while this PR uses "projection shortcuts only when dimensions differ".

Comment (RecreationalMath) [reply sktime/networks/resnet.py:93]: Upon further review, this is the correct way to generalize the currently hard-coded layers. Referenced papers wang/hfawaz don't mention projections (perhaps) because it is an implementational detail, while the comments in [hfawaz/dl-4-tsc](https://github.com/hfawaz/dl-4-tsc/blob/e0233efd886df8c6ca18e6f1b545d15aaf423627/classifiers/resnet.py#L75-L96) mention the intention behind them as channel mismatch. 

No action needed. This is correct.

File: sktime/networks/resnet.py [hunk 6]
```diff
@@ -60,94 +78,35 @@ def build_network(self, input_shape, **kwargs):
...
-        output_block_3 = keras.layers.Activation(self.activation)(output_block_3)
+        x = input_layer
+
+        # stack residual blocks; number of blocks = len(self.n_filters)
+        for filters in self.n_filters:
+
+            out = x
+            for i, kernel_size in enumerate(self.kernel_sizes):
+                out = keras.layers.Conv1D(
+                    filters=filters, kernel_size=kernel_size, padding=self.padding
```
Comment (RecreationalMath) [inline sktime/networks/resnet.py]: After reading the referenced papers and reviewing the implementation in [hfawaz/dl-4-tsc](https://github.com/hfawaz/dl-4-tsc/blob/master/classifiers/resnet.py), I realized that using any padding other than `"same"` will break the architecture.

- The convolution path goes through three Conv1D layers (kernels 8, 5, and 3 by default). With `"valid"` padding, each layer shrinks the sequence length. After 3 layers: length shrinks by (8-1) + (5-1) + (3-1) = 14
- The shortcut path applies either BatchNorm (no length change) or Conv1D(kernel=1), which also preserves length
- When the outputs are added (`keras.layers.add([shortcut, out])`), the shapes no longer match, leading to a runtime error.

So in this architecture, padding effectively needs to be `"same"`, which makes exposing it as a configurable parameter misleading.
Apologies for my earlier suggestion, that was my oversight. 

In general, it’s always worth double-checking review suggestions (including mine), since subtle architectural constraints like this are easy to miss.

Comment (Rusheel86) [reply sktime/networks/resnet.py]: Thanks for the thorough review, I appreciate you catching that. I've removed padding as an exposed parameter and hardcoded it back to "same". Please let me know if anything else needs addressing.

**Comment by Rusheel86** [comment]
@RecreationalMath , sorry for the ping. The pr is ready for review whenever you have time.

**Comment by Rusheel86** [comment]
Its been a while can someone review this pr please?
@RecreationalMath  @fkiraly
